docker/indexer/stages/preparation/preparation.go

```diff
@@ -0,0 +1,308 @@
+type Checker interface {
+	Exists(ctx context.Context, name string, hash plumbing.Hash) (bool, error)
+}
+
+// Stage holds the data types necessary to process repository configuration.
+type Stage struct {
+	Checker Checker
+	RepoHdl *storage.BucketHandle
+}
+
+// Run runs the stage and outputs Result data types to the results channel.
+func (s *Stage) Run(ctx context.Context, cfgs []*config.RepoConfig, results chan *Result) error {
+	var err error
+	wErr := make(chan error, workers)
+
+	wCtx, wCancel := context.WithCancel(ctx)
+	defer wCancel()
+
+	sem := semaphore.NewWeighted(workers)
+	for _, repoCfg := range cfgs {
+		if err := sem.Acquire(wCtx, 1); err != nil {
+			return fmt.Errorf("failed to acquire semaphore: %v", err)
+		}
+
+		go func(ctx context.Context, repoCfg *config.RepoConfig, results chan *Result, errCh chan error) {
+			defer sem.Release(1)
+
+			var err error
+			select {
+			case <-ctx.Done():
+				errCh <- context.Canceled
+				return
+			default:
+			}
+			log.Infof("received config for %s", repoCfg.Name)
+			switch repoCfg.Type {
+			case shared.Git:
+				err = s.processGit(ctx, repoCfg, results)
+			default:
+				errCh <- fmt.Errorf("unsupported repository type %s", repoCfg.Type)
+			}
+			if err != nil {
+				wErr <- err
+			}
+		}(wCtx, repoCfg, results, wErr)
+
+		select {
+		case err = <-wErr:
+			log.Errorf("preparation worker returned an error: %v", err)
+		default:
+		}
+	}
+	return sem.Acquire(ctx, workers)
+}
+
+func (s *Stage) objectExists(ctx context.Context, name string) bool {
+	objItr := s.RepoHdl.Objects(ctx, &storage.Query{Prefix: name + shared.TarExt})
+	_, err := objItr.Next()
+	return err == nil
+}
+
+func (s *Stage) processGit(ctx context.Context, repoCfg *config.RepoConfig, results chan *Result) error {
+	var (
+		err     error
+		repo    *git.Repository
+		repoDir string
+	)
+	if !s.objectExists(ctx, repoCfg.Name) {
+		repo, repoDir, err = s.cloneGitRepo(ctx, repoCfg.Name, repoCfg.Address)
+	} else {
+		repo, repoDir, err = s.updateGitRepo(ctx, repoCfg.Name)
+	}
+	if repoDir != "" {
+		defer func() {
+			if err := os.RemoveAll(repoDir); err != nil {
+				log.Errorf("failed to remove local repo: %v", err)
+			}
+		}()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	comItr, err := repo.CommitObjects()
+	if err != nil {
+		return err
+	}
+	allCommits := make(map[plumbing.Hash]*object.Commit)
+	comItr.ForEach(func(c *object.Commit) error {
+		allCommits[c.Hash] = c
+		return nil
+	})
+
+	commitTracker := make(map[plumbing.Hash]bool)
+	// repoInfo is used as the iterator function to create RepositoryInformation structs.
+	repoInfo := func(ref *plumbing.Reference) error {
+		found, err := s.Checker.Exists(ctx, repoCfg.Name, ref.Hash())
+		if err != nil {
+			return err
+		}
+		if found {
+			return nil
+		}
+
+		var when time.Time
+		if c, ok := allCommits[ref.Hash()]; ok {
+			when = c.Author.When
+		}
+		versRE, err := regexp.Compile(repoCfg.VersionRE)
```

Comment:
Can we make VersionRE optional? As this an unspecified VersionRE will result in an error right?